File: hot100/Solution0019.cpp
#include <stack>
using namespace std;
// 给你一个链表，删除链表的倒数第 n 个结点，并且返回链表的头结点。

// 示例 1：

// 输入：head = [1,2,3,4,5], n = 2
// 输出：[1,2,3,5]
// 示例 2：

// 输入：head = [1], n = 1
// 输出：[]
// 示例 3：

// 输入：head = [1,2], n = 1
// 输出：[1]

// 提示：

// 链表中结点的数目为 sz
// 1 <= sz <= 30
// 0 <= Node.val <= 100
// 1 <= n <= sz

// 进阶：你能尝试使用一趟扫描实现吗？
class Solution
{
private:
    struct ListNode
    {
        int val;
        ListNode *next;
        ListNode(int x) : val(x), next(nullptr) {}
        ListNode(int x, ListNode *next) : val(x), next(next) {}
    };

public:
    ListNode *removeNthFromEnd(ListNode *head, int n)
    {
        // 快慢指针：快指针先走n步。如果第n步时，fast_pointer.next == nullptr，说明删除的是头节点，返回head.next即可
        ListNode *fast_pointer = head, *slow_pointer = head;
        for (int i = 0; i < n; ++i)
        {
            fast_pointer = fast_pointer->next;
        }
        if (!fast_pointer)
            return head->next;
        while (fast_pointer->next != nullptr)
        {
            fast_pointer = fast_pointer->next;
            slow_pointer = slow_pointer->next;
        }
        slow_pointer->next = slow_pointer->next->next;
        return head;
    }
};

// 方法二：栈
class Solution2
{
private:
    struct ListNode
    {
        int val;
        ListNode *next;
        ListNode(int x) : val(x), next(nullptr) {}
        ListNode(int x, ListNode *next) : val(x), next(next) {}
    };

public:
    ListNode *removeNthFromEnd(ListNode *head, int n)
    {
        stack<ListNode *> nodeStack;
        ListNode *dummy = new ListNode(0, head);
        ListNode *current = dummy;
        while (current != nullptr)
        {
            nodeStack.push(current);
            current = current->next;
        }
        for (int i = 0; i < n; ++i)
        {
            nodeStack.pop();
        }
        ListNode *prev = nodeStack.top();
        ListNode *toDelete = prev->next;
        prev->next = toDelete->next;
        delete toDelete;
        return dummy->next;
    }
};

File: hot100/Solution0019.java
package hot100;

import java.util.ArrayDeque;
import java.util.Deque;

public class Solution0019 {
    private static class ListNode {
        int val;
        ListNode next;

        ListNode(int x) {
            val = x;
            next = null;
        }
    }

    public ListNode removeNthFromEnd(ListNode head, int n) {
        ListNode slow = head;
        ListNode fast = head;
        for (int i = 0; i < n; i++) {
            fast = fast.next;
        }
        if (fast == null) {
            return head.next;
        }
        while (fast.next != null) {
            slow = slow.next;
            fast = fast.next;
        }
        slow.next = slow.next.next;
        return head;
    }

    public ListNode removeNthFromEnd2(ListNode head, int n) {
        Deque<ListNode> stack = new ArrayDeque<>();
        ListNode current = head;
        while (current != null) {
            stack.push(current);
            current = current.next;
        }
        for (int i = 0; i < n; i++) {
            stack.pop();
        }
        if (stack.isEmpty()) {
            return head.next;
        }
        ListNode prev = stack.peek();
        prev.next = prev.next.next;
        return head;
    }

    public static void main(String[] args) {
        Solution0019 solution = new Solution0019();
        ListNode head = new ListNode(1);
        head.next = new ListNode(2);
        head.next.next = new ListNode(3);
        head.next.next.next = new ListNode(4);
        head.next.next.next.next = new ListNode(5);

        int n = 2;
        ListNode result = solution.removeNthFromEnd(head, n);
        System.out.print("删除倒数第 " + n + " 个节点后的链表: ");
        while (result != null) {
            System.out.print(result.val + " ");
            result = result.next;
        }
    }
}
